tensorflow_addons/layers/multihead_attention_test.py

```diff
@@ -0,0 +1,338 @@
+
+        mha = MultiHeadAttention(
+            head_size=head_size, num_heads=num_heads, return_attn_coef=True
+        )
+
+        output, attn_coef = mha([q, k, v], mask=mask)
+
+        self.assertEqual(output.shape[0], batch_size)
+        self.assertEqual(output.shape[1], q.shape[1])
+        self.assertEqual(output.shape[2], v.shape[2])
+
+        if tf.executing_eagerly():
+            attn_coef = attn_coef.numpy()
+            mask = mask.numpy()
+
+            self.assertTrue(((attn_coef != 0) == mask).all())
+
+    def test_from_to_config(self):
+        num_heads = 8
+        head_size = 12
+
+        mha = MultiHeadAttention(head_size=head_size, num_heads=num_heads, dropout=0.5)
+
+        config = mha.get_config()
+
+        new_mha = MultiHeadAttention.from_config(config)
+
+        self.assertEqual(mha.head_size, new_mha.head_size)
+        self.assertEqual(mha.num_heads, new_mha.num_heads)
+        self.assertEqual(mha._droput_rate, new_mha._droput_rate)
+
+    def test_save_load_model(self):
+
+        num_heads = 8
+        head_size = 12
+
+        inputs = tf.keras.layers.Input(shape=[42, 13])
+
+        net, attn_coef = MultiHeadAttention(
+            head_size=head_size, num_heads=num_heads, dropout=0.5, return_attn_coef=True
+        )([inputs, inputs, inputs])
+        net = tf.keras.layers.GlobalAveragePooling1D()(net)
+        net = tf.keras.layers.Dense(10, activation="softmax")(net)
+
+        model = tf.keras.Model(inputs=inputs, outputs=[net, attn_coef])
+
+        # initialize model
+        model.predict(np.random.uniform(size=(10, 42, 13)))
+
+        with tempfile.TemporaryDirectory() as model_dir:
+            model_path = str(Path(model_dir) / "saved_model")
+            model.save(model_path)
+            new_model = tf.keras.models.load_model(model_path)
+
+        self.assertEqual(model.layers[1].get_config(), new_model.layers[1].get_config())
+
+    def test_fit_predict_eval(self):
+
+        num_heads = 8
+        head_size = 12
+
+        inputs = tf.keras.layers.Input(shape=[42, 13])
+
+        net = MultiHeadAttention(head_size=head_size, num_heads=num_heads, dropout=0.5)(
+            [inputs, inputs, inputs]
+        )
+        net = tf.keras.layers.GlobalAveragePooling1D()(net)
+        net = tf.keras.layers.Dense(10, activation="softmax")(net)
+
+        model = tf.keras.Model(inputs=inputs, outputs=net)
+
+        model.compile(
+            loss=tf.losses.SparseCategoricalCrossentropy(),
+            optimizer=tf.keras.optimizers.Adam(0.001),
+        )
+
+        model.fit(
+            x=np.random.uniform(size=(50, 42, 13)),
+            y=np.random.randint(10, size=(50,)),
+            batch_size=10,
+            epochs=2,
+        )
+
+        model.predict(np.random.uniform(size=(10, 42, 13)))
+
+        model.evaluate(
+            x=np.random.uniform(size=(20, 42, 13)),
+            y=np.random.randint(0, 10, size=(20,)),
+            batch_size=10,
+        )
+
+
```

Comment:
Could you add a simple test that uses keras.layer_test:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/utils/test_utils.py#L32

Similar to:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/sparsemax_test.py#L59

This utility verifies a number of things (such as ser/deser) and we'd like to use uniform testing for subpackages as much as possible. See #992

Reply:
@seanpmorgan the `layer_test` docs say its a

> Test routine for a layer with a single input and single output.

Not sure if I can use it because `MutiHeadAttention` takes 3 inputs.

Reply:
Ah, you're correct no need then and you've implemented the serialization tests outside of this anyway.